core/src/main/scala/org/bitcoins/core/protocol/ln/LnTaggedFields.scala

```
@@ -0,0 +1,199 @@
+    fallbackAddress: Option[LnTag.FallbackAddressTag] = None,
+    routingInfo: Option[LnTag.RoutingInfo] = None): LnTaggedFields = {
+
+    if (descriptionOrHash.isLeft) {
+      InvoiceTagImpl(
+        paymentHash = paymentHash,
+        description = descriptionOrHash.left.toOption,
+        nodeId = nodeId,
+        descriptionHash = None,
+        expiryTime = expiryTime,
+        cltvExpiry = cltvExpiry,
+        fallbackAddress = fallbackAddress,
+        routingInfo = routingInfo)
+    } else {
+
+      InvoiceTagImpl(
+        paymentHash = paymentHash,
+        description = None,
+        nodeId = nodeId,
+        descriptionHash = descriptionOrHash.right.toOption,
+        expiryTime = expiryTime,
+        cltvExpiry = cltvExpiry,
+        fallbackAddress = fallbackAddress,
+        routingInfo = routingInfo)
+    }
+
+  }
+
+  def fromUInt5s(u5s: Vector[UInt5]): LnTaggedFields = {
+    @tailrec
+    def loop(remaining: List[UInt5], fields: Vector[LnTag]): Vector[LnTag] = {
+      remaining match {
+        case h :: h1 :: h2 :: t =>
+
+          val prefix = LnTagPrefix.fromUInt5(h)
+
+          //next two 5 bit increments are data_length
+          val dataLengthU5s = Vector(h1, h2)
+
+          val dataLength = LnUtil.decodeDataLength(dataLengthU5s)
+
+          //t is the actual possible payload
+          val payload: Vector[UInt5] = t.take(dataLength.toInt).toVector
+
+          val tag = LnTag.fromLnTagPrefix(prefix.get, payload)
+
+          val newRemaining = t.slice(payload.size, t.size)
+
+          loop(newRemaining, fields.:+(tag))
+
+        case _ :: _ | _ :: _ :: _ =>
+          throw new IllegalArgumentException("Failed to parse LnTaggedFields, needs 15bits of meta data to be able to parse")
+        case Nil =>
+          fields
+      }
+    }
+
+    val tags = loop(u5s.toList, Vector.empty)
+
+    val paymentHashTag = tags.find(_.isInstanceOf[LnTag.PaymentHashTag])
+      .get.asInstanceOf[LnTag.PaymentHashTag]
```

`PaymentHashTag` is required, but we should probably throw a nice exception or return a `Try` if it is not found